cluster_tools/cluster_tools/executors/dask.py

```diff
@@ -28,23 +33,103 @@
 _S = TypeVar("_S")
 
 
+def _run_in_nanny(
+    queue: Queue, __fn: Callable[_P, _T], *args: _P.args, **kwargs: _P.kwargs
+) -> None:
+    try:
+        __env = cast(Dict[str, str], kwargs.pop("__env"))
+        for key, value in __env.items():
+            os.environ[key] = value
+
+        ret = __fn(*args, **kwargs)
+        queue.put({"value": ret})
+    except Exception as exc:
+        queue.put({"exception": exc})
+
+
+def _run_with_nanny(
+    __fn: Callable[_P, _T],
+    *args: _P.args,
+    **kwargs: _P.kwargs,
+) -> _T:
+    mp_context = get_context("spawn")
+    q = mp_context.Queue()
+    p = mp_context.Process(target=_run_in_nanny, args=(q, __fn) + args, kwargs=kwargs)
+    p.start()
+    p.join()
+    ret = q.get(timeout=0.1)
+    if "exception" in ret:
+        raise ret["exception"]
+    else:
+        return ret["value"]
+
+
+def _parse_mem(size: str) -> int:
+    units = {"": 1, "K": 2**10, "M": 2**20, "G": 2**30, "T": 2**40}
+    m = re.match(r"^([\d\.]+)\s*([a-zA-Z]{0,3})$", str(size).strip())
+    assert m is not None
+    number, unit = float(m.group(1)), m.group(2).upper()
+    return int(number * units[unit])
+
+
 class DaskExecutor(futures.Executor):
+    """
+    The `DaskExecutor` allows to run workloads on a dask cluster.
+
+    The executor can be constructed with an existing dask `Client` or
+    from a declarative configuration. The address of the dask scheduler
+    can be part of the configuration or supplied as environment variable
+    `DASK_ADDRESS`.
+
+    There is support for resource-based scheduling. As default, `mem` and
+    `cpus-per-task` are supported. To make use of them, the dask workers
+    should be started with:
+    `python -m dask worker --no-nanny --nthreads 6 tcp://... --resources "mem=1073741824 cpus=8"`
+    """
+
     client: "Client"
+    pending_futures: Set[Future]
+    job_resources: Optional[Dict[str, Any]]
+    is_shutting_down = False
 
     def __init__(
-        self,
-        client: "Client",
+        self, client: "Client", job_resources: Optional[Dict[str, Any]] = None
     ) -> None:
         self.client = client
+        self.pending_futures = set()
+        self.job_resources = job_resources
+
+        if self.job_resources is not None:
+            # `mem` needs to be a number for dask, so we need to parse it
+            if "mem" in self.job_resources:
+                self.job_resources["mem"] = _parse_mem(self.job_resources["mem"])
+            if "cpus-per-task" in self.job_resources:
+                self.job_resources["cpus"] = int(
+                    self.job_resources.pop("cpus-per-task")
+                )
+
+        # Clean up if a SIGINT signal is received. However, do not interfere with the
```

Comment:
This is copied from `ClusterExecutor`. It seems that the SIGINT handler is never unattached. I would've expected that to happen in `shutdown`.

cc @daniel-wer

Reply:
Hmm, I don't remember why I didn't add that. However, thinking about it, this could get tricky. Assuming the signal handler might have been changed in the mean time, I'm not sure how a proper cleanup would look like or is possible at all :thinking:

Reply:
Maybe we should use a weakref to avoid a larger memory leak through the executor.

Reply:
Clever solution :+1: